engelsystem/src/main/java/info/metadude/android/eventfahrplan/engelsystem/EngelsystemNetworkRepository.kt

```diff
@@ -1,26 +1,25 @@
 package info.metadude.android.eventfahrplan.engelsystem
 
-import info.metadude.android.eventfahrplan.engelsystem.loading.ShiftsLoading.awaitShiftsResult
 import info.metadude.android.eventfahrplan.engelsystem.models.ShiftsResult
 import info.metadude.android.eventfahrplan.engelsystem.utils.UriParser
-import info.metadude.kotlin.library.engelsystem.ApiModule
 import okhttp3.OkHttpClient
-import java.net.URISyntaxException
+import retrofit2.HttpException
 
 class EngelsystemNetworkRepository(
 
-        private val uriParser: UriParser = UriParser()
+    private val uriParser: UriParser = UriParser(),
+    private val serviceProvider: EngelsystemServiceProvider = EngelsystemServiceProvider.getNewInstance()
 
 ) {
 
     suspend fun load(okHttpClient: OkHttpClient, url: String) = try {
         val uri = uriParser.parseUri(url)
-        val service = ApiModule.provideEngelsystemService(uri.baseUrl, okHttpClient)
-        val call = service.getShifts(uri.pathPart, uri.apiKey)
-        call.awaitShiftsResult()
-    } catch (e: URISyntaxException) {
-        ShiftsResult.Exception(e)
-    } catch (e: IllegalArgumentException) {
+        val service = serviceProvider.getService(uri.baseUrl, okHttpClient)
+        val shifts = service.getShifts(uri.pathPart, uri.apiKey)
+        ShiftsResult.Success(shifts)
+    } catch (e: HttpException) {
+        ShiftsResult.Error(e.code(), e.message())
+    } catch (e: Exception) {
```

Comment:
Maybe you should add another case for Corotines `CancellationException`s, because one should re-throw them (I hope this is still true, otherwise ignore it :) ): https://www.lukaslechner.com/7-common-mistakes-you-might-be-making-when-using-kotlin-coroutines/

Reply:
I am not sure if `ShiftsResult.Exception(e)` handles this CancellationException correctly, or if a re-throw is necessary.

Reply:
Thank you for sharing the article. I tried out the same pattern as described in _common mistake 7_. When I cancelled the job after 500ms I got the following output:

```
         System.out  I  Performing network request
                     I  Catching exception
StandaloneCoroutine  E  ShiftsResult.Exception: StandaloneCoroutine was cancelled
         System.err  W  kotlinx.coroutines.JobCancellationException: StandaloneCoroutine was cancelled; job=StandaloneCoroutine{Cancelling}@fc81a91
```

No output that the coroutine is still running. It looks good to me. Did you observe a different behavior?

Reply:
Seems fine. No, I did not observe something different, it was just in the back of my mind.